engine/src/main/java/org/hibernate/search/engine/cfg/spi/NormalizeUtils.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.search.engine.cfg.spi;
+
+import java.time.Instant;
+import java.time.OffsetTime;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+
+import org.hibernate.search.util.common.annotation.Incubating;
+
+@Incubating
+public final class NormalizeUtils {
+
+	private static final long ALMOST_WHOLE_DAY = 3600_000_000_000L * 24 - 1;
+
+	private NormalizeUtils() {
+	}
+
+	public static OffsetTime normalizeOffsetTime(Long encoded) {
+		Instant instant = Instant.EPOCH.plus( encoded, ChronoUnit.NANOS );
+		if ( encoded < 0 ) {
+			return instant.atOffset( ZoneOffset.ofHours( 24 - instant.atOffset( ZoneOffset.UTC ).getHour() ) )
+					.toOffsetTime();
+		}
+
+		if ( encoded > ALMOST_WHOLE_DAY ) {
+			return instant.atOffset( ZoneOffset.ofHours( -instant.atOffset( ZoneOffset.UTC ).getHour() - 1 ) )
+					.toOffsetTime();
+		}
+
+		return instant.atOffset( ZoneOffset.UTC ).toOffsetTime();
+	}
```

Comment:
hey @yrodiere  👋🏻 🙈 

I wanted to ask you what you'd think about such "timzeone-offset-magic" 🪄🪄🪄... 

All this is to make using `RAW` on input work fine... we are decoding and then encoding the values through a codec. If a user passes us a long-as-offset-time we'll end up losing the fact that e.g. the time is in the previous/next day... e.g. 

Let's assume we have a `17-00+18`, this would encode to -1hour (in nanos). Decoding that will result in `23:01Z`. 

This means if we had a range of `( 17-00+18 ; 19-00+18 )`, which seems legit, if such range goes through encode decode, currently it'll result in `( 23:01Z  ; 01:01Z)` and here if we encode the lower and upper values again, we'll have `lower > upper` leading to an exception (`82860000000000 ; 3660000000000`). With the change ... we'll get `(00:01+01:00 ; 01:01Z)` or, if encoded (`-3540000000000 ; 3660000000000`), preserving the "order". 

The alternative is to bypass the dsl converters entirely, but that means we'll have something like:
```java
if (valueModel == RAW) {
    if (codec.getEncodedType() != compatible with requested type) {
        throw log.someExcetpion();
    }

    return value;
} else {
    var converter = getDslConverter();

    codec.encode(converter.covert(value)
}
```
and I've tried that & but didn't like it 🙈 😕... + that won't work fine with ES backend as there we didn't want to expose JSON types to the user ...

Reply:
> The alternative is to bypass the dsl converters entirely

That's actually what I'd expect. I mean, raw is raw. No conversion, pass what the backend needs directly.

> and I've tried that & but didn't like it 🙈 😕... 

Can't argue about feelings :shrug: 

> + that won't work fine with ES backend as there we didn't want to expose JSON types to the user ...

What about expecting whatever we feed into JSON types? E.g. expect a String if we're using `new JsonPrimitive(<String>)`.

Yes it's brutally simple and potentially brittle, but then again, it _is_ called "raw".

Reply:
> > * that won't work fine with ES backend as there we didn't want to expose JSON types to the user ...
> 
> What about expecting whatever we feed into JSON types? E.g. expect a String if we're using `new JsonPrimitive(<String>)`.
> 
> Yes it's brutally simple and potentially brittle, but then again, it _is_ called "raw".

Or you could expect a JSON string and parse it, if that makes life easier for you (e.g. for native field types).

Reply:
> Or you could expect a JSON string and parse it, if that makes life easier for you (e.g. for native field types).

That's what I've essentially done for now (at least I think so 😄). I'm using a parser converter to parse a string that the user passed through the DSL to us. The parser should be able to parse strings in the format that ES expects, so that seems to work okay.

Reply:
> Yes it's brutally simple and potentially brittle, but then again, it is called "raw".
> Can't argue about feelings 🤷

ok, you've convinced me 😃. I'll give it another try. There was a nice rainstorm over the weekend and maybe it changed my feelings towards it 😄😄😄

Reply:
I've been looking at this ... and it seems that to make it work and to keep it clean, we'd need to move the codec encode/decode into dsl/projection converter ... does that sound like a good idea to you?

Reply:
> to keep it clean, we'd need to move the codec encode/decode into dsl/projection converter ...

ok, so this idea didn't work out 😖😖😖 elasticsearch codecs have multiple encode functions for various parts so having encode/decode packed inside the dsl/projection converter. I'll try a straightforward approach to just check things in place. I'll ping you once I have something workable.